### ISSUE TITLE
lint:missing-argument when splatting into a function that has a compatible positional argument.

### ISSUE THREAD

**Issue by dreid** [astral-sh/ty#247]
### Summary

Minimal reproducer: https://play.ty.dev/4b8dc795-03b5-4fbf-a5c3-639f9518e093

I'd expect first the invocation to type check because the splatted args have compatible types to cover both the first argument and the var args.

**Comment by carljm** [comment]
Thanks for the report! Yeah, we don't support splatted arguments yet, but we should fix that soon.

**Comment by GriceTurrble** [comment]
Similar scenario to watch for, which brought me here:

https://play.ty.dev/edb289aa-d398-4f51-b758-42f584ca80e1

```py
def something(foo, **others):
    pass

kwargs = {
    "foo": "one",
    "bar": "two",
}
something(**kwargs)
```

This throws the error:
```
No argument provided for required parameter `foo` of function `something` (missing-argument) [Ln 8, Col 1]
```

`foo` is part of the kwargs dict provided, but it's not recognized as being present.

**Comment by spapanik** [comment]
I am not sure if this is the same issue, I guess the root cause is the same. I am getting this error:

```py
error[unresolved-attribute]: Type `str` has no attribute `pop`
   --> src/eulertools/lib/utils.py:383:18
    |
381 |     prefix, response_key, *answers = line.split(maxsplit=2)
382 |     try:
383 |         answer = answers.pop()
    |                  ^^^^^^^^^^^
384 |     except IndexError:
385 |         answer = ""
    |
info: rule `unresolved-attribute` is enabled by default
```

**Comment by carljm** [comment]
I think the most recent example here is actually https://github.com/astral-sh/ty/issues/191 -- the other examples do look like this issue.

**Comment by dhruvmanila** [comment]
@spapanik The `unresolved-attribute` issue mentioned in https://github.com/astral-sh/ty/issues/247#issuecomment-2917656676 should be fixed by astral-sh/ruff#18438

**Comment by Andrej730** [comment]
A note that currently other type checkers (pyright, mypy) do not check very precisely some cases of this and would allow errors like below to pass. So probably it would be okay for ty not to cover them too for a while.

```python
def test(a: int, b: int): ...

test(**{"a": 25, "b": 42, "c": 11})
test(*[5, 4, 23])
```

**Comment by carljm** [comment]
Yes, I think that's perhaps not only a "for now" thing, but possibly a "for always" that we will have to match mypy/pyright's forgiving handling of splatted calls into non-variadic signatures, because I fear the backwards-compatibility impact would otherwise be too great. Perhaps at some point in the future we could explore more opt-in strictness here, but that's definitely not an immediate priority.

**Comment by sharkdp** [comment]
@dcreager I understand that this is not yet finished, despite https://github.com/astral-sh/ruff/pull/18996 being merged? Do you think we can remove the "Missing support for starred/splatted/unpacked arguments in function calls" item from https://github.com/astral-sh/ty/issues/445 (after we cut a release)?

**Comment by dcreager** [comment]
It should now work for calls to non-overloaded functions, but will likely produce invalid type inferences and diagnostics for calls to overloaded functions.

**Comment by dcreager** [comment]
e.g. I updated the snippet to use tuples instead of lists (since we still aren't inferring a proper type for list literals), and it works as expected [[playground](https://play.ty.dev/f55d3567-da1f-4a0d-ac26-ae6963e4043b)]:

```py
def foo(initial_arg: str, *args: str) -> None:
    ...

foo(*("a", "b", "c"))

# error: [invalid-argument-type]
foo(*(1, "a", "b"))
```

**Comment by hauntsaninja** [comment]
I would consider this a false positive on a non-overloaded function: https://play.ty.dev/92958251-1a2e-4238-b77c-83983bd7706f (yes, there can be errors here, but mypy and pyright accept it and this kind of thing is pretty common)

**Comment by dcreager** [comment]
Ah sorry — I forgot to call out that https://github.com/astral-sh/ruff/pull/18996 only added support for positional splats, not keyword splats.

**Comment by sharkdp** [comment]
This has been closed, but the example given in https://github.com/astral-sh/ty/issues/445 still leads to errors (diagnostic code has changed since writing that example). This is the current behavior:
```py
def f(a: str, b: int): pass

opts = {"name": "foo", "n": 2}

# invalid-argument-type: Argument to function `f` is incorrect: Expected `str`, found `Unknown | str | int`
# invalid-argument-type: Argument to function `f` is incorrect: Expected `int`, found `Unknown | str | int`
f(**opts)
```

Is that being tracked somewhere else?

**Comment by carljm** [comment]
I think that's what #1248 is about?

No other type checker really "supports" that example in any intelligent way. Pyright (in non-strict mode) avoids false positives by inferring the type of the dict as `dict[str, Unknown]`. Strict-mode pyright, pyrefly, and mypy all fail the same way we do.

**Comment by AlexWaygood** [comment]
theoretically I guess you could bidirectionally infer the type of `opts` as being a synthesized `TypedDict` type due to the fact that it's kwarg-splatted into the `f()` call later on in the same scope. Sounds kinda hard though
